### PR TITLE
Multitarget to .net core 3.1 and .net 5.0

### LINKED ISSUE #24: Target net5.0
At the moment referencing package in net5.0 WPF app I get the following warning:

**Package 'Emoji.Wpf 0.1.5' was restored using '.NETFramework,Version=v4.6.1, .NETFramework,Version=v4.6.2, .NETFramework,Version=v4.7, .NETFramework,Version=v4.7.1, .NETFramework,Version=v4.7.2, .NETFramework,Version=v4.8' instead of the project target framework 'net5.0-windows7.0'. This package may not be fully compatible with your project.**

Though it works without an issue, it would be nice to target net5.0 as well.
Thanks.

### PR DESCRIPTION
**Pull request by virzak**
Fixes #24

### CODE REVIEW

File: Emoji.Wpf/Emoji.Wpf.csproj [hunk 1]
```diff
@@ -21,111 +17,53 @@
...
-    <AppDesignerFolder>Properties</AppDesignerFolder>
-    <RootNamespace>Emoji.Wpf</RootNamespace>
-    <AssemblyName>Emoji.Wpf</AssemblyName>
-    <TargetFrameworkVersion>v4.0</TargetFrameworkVersion>
-    <FileAlignment>512</FileAlignment>
-    <TargetFrameworkProfile />
+    <TargetFrameworks>net40;netcoreapp3.1;net50-windows</TargetFrameworks>
+    <GenerateAssemblyFileVersionAttribute>false</GenerateAssemblyFileVersionAttribute>
+  <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
+  <GenerateAssemblyVersionAttribute>false</GenerateAssemblyVersionAttribute>
```
Comment (drewnoakes) [inline Emoji.Wpf/Emoji.Wpf.csproj:28]: Why these properties? The `AssemblyInfo.cs` file has been removed. Usually you set these when you want to use attributes in that file instead.

File: Emoji.Wpf/Emoji.Wpf.csproj [hunk 2]
```diff
@@ -21,111 +17,53 @@
...
-      <Generator>MSBuild:Compile</Generator>
-      <SubType>Designer</SubType>
-    </Page>
-    <Page Include="TextBox.xaml">
-      <SubType>Designer</SubType>
-      <Generator>MSBuild:Compile</Generator>
-    </Page>
+    <None Remove="WinFonts\webdings.txt" />
+    <None Remove="WinFonts\wingdings.txt" />
+    <None Remove="$(OutputPath)\emoji-test.txt.gz" />
```
Comment (drewnoakes) [inline Emoji.Wpf/Emoji.Wpf.csproj:44]: These removals should not be needed.

File: Emoji.Wpf/Emoji.Wpf.csproj [hunk 3]
```diff
@@ -1,9 +1,5 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props"
-          Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
-  <UsingTask TaskName="GZipTask" TaskFactory="CodeTaskFactory"
-             AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll">
+﻿<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
```
Comment (drewnoakes) [inline Emoji.Wpf/Emoji.Wpf.csproj:1]: I don't believe the `WindowsDesktop` SDK is needed any more.

```suggestion
﻿<Project Sdk="Microsoft.NET.Sdk">
```

File: FontViewer/FontViewer.csproj [hunk 4]
```diff
@@ -1,69 +1,10 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+﻿<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
```
Comment (drewnoakes) [inline FontViewer/FontViewer.csproj:1]: Ditto, re `WindowsDesktop`.

File: Sample/Sample.csproj [hunk 5]
```diff
@@ -1,69 +1,10 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+﻿<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
```
Comment (drewnoakes) [inline Sample/Sample.csproj:1]: Ditto, re `WindowsDesktop`.

File: Tests/Tests.csproj [hunk 6]
```diff
@@ -1,69 +1,10 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+﻿<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
```
Comment (drewnoakes) [inline Tests/Tests.csproj:1]: Ditto, re `WindowsDesktop`.

File: Tiny/Tiny.csproj [hunk 7]
```diff
@@ -1,69 +1,10 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+﻿<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
```
Comment (drewnoakes) [inline Tiny/Tiny.csproj:1]: Ditto, re `WindowsDesktop`.

**Comment by samhocevar** [comment]
Thanks for this PR. There is too much noise in these changes for me to merge it as is, but I think I can cherry-pick the relevant changes.

File: Emoji.Wpf/Emoji.Wpf.csproj (hunk 1 above)
Comment (samhocevar) [reply Emoji.Wpf/Emoji.Wpf.csproj:28]: This is necessary because these attributes are already generated by GitVersion.

**Comment by drewnoakes** [comment]
@samhocevar I'm curious what you mean by noise? You'll have to convert the projects to SDK-style in order to successfully multi-target. This PR seems like an effective conversion to me (with the caveat that I've only reviewed the change on GitHub). I work on the Visual Studio component that deals with SDK-style projects, which is why I took an interest in this PR and your appraisal of it.

File: Emoji.Wpf/Emoji.Wpf.csproj (hunk 1 above)
Comment (drewnoakes) [reply Emoji.Wpf/Emoji.Wpf.csproj:28]: I would expect GitVersion to do this suppression, in that case.

**Comment by samhocevar** [comment]
@drewnoakes sorry, I should have clarified what I meant. These changes are good and helpful, but I would like to make minor adjustments before I merge them.
 - the `GZipTask` definition and usage have been reformatted
 - there are a few indentation errors inside the `<PropertyGroup>` node
 - as it is, opening `Emoji.Wpf.csproj` inside Visual Studio will show the entire directory structure of the `UnicodeCLDR` and `UnicodeEmoji` submodules (only one file is actually used)
 - at least one change seems unrelated to the intent of the PR (upgrading the GitVersion package from 5.6.4 to 5.6.6)

All of this is harmless, just not very clean IMHO.

**Comment by drewnoakes** [comment]
@samhocevar thanks for clarifying. Perhaps give that feedback to @virzak who clearly went to a lot of effort to submit this, and give them the opportunity to have this merged. Your project though.

File: Emoji.Wpf/Emoji.Wpf.csproj (hunk 3 above)
Comment (virzak) [reply Emoji.Wpf/Emoji.Wpf.csproj]: Just tried it but unfortunately it breaks net40 target, so had to revert.

Comment (virzak) [reply Emoji.Wpf/Emoji.Wpf.csproj]: @drewnoakes, I think this is the issue: https://github.com/dotnet/sdk/issues/15475

File: Emoji.Wpf/Emoji.Wpf.csproj (hunk 1 above)
Comment (virzak) [reply Emoji.Wpf/Emoji.Wpf.csproj]: GitVersion was conflicting, so these lines had to be there

File: Tests/Tests.csproj (hunk 6 above)
Comment (virzak) [reply Tests/Tests.csproj:1]: Same issue (net40)

**Comment by virzak** [comment]
@samhocevar All the adjustments you've requested are in apart from going back to 5.6.4.

As the SDK style requires a complete rewrite of the old project files, the best way to go about it is to start fresh and only add what's needed either by hand or through Visual Studio UI.

The formatting got affected because Visual Studio reformats as it pleases, so I suggest adding an `.editoconfig` file and setting the proper indentation, otherwise it is easy to mess up.

Since this is a rewrite, I didn't see a harm in getting the latest version of the  GitVersion package, but if you'd like we can go back.

File: Emoji.Wpf/Emoji.Wpf.csproj (hunk 3 above)
Comment (samhocevar) [reply Emoji.Wpf/Emoji.Wpf.csproj]: I made some tests the project does not build without it, seemingly because of `<UseWPF>true</UseWPF>`.

**Comment by virzak** [comment]
I removed net50-windows since there is not code difference between netcoreapp3.1 and net50-windows

The guidance states:

```
❌ AVOID multi-targeting as well as targeting .NET Standard, if your source code is the same for all targets.

The .NET Standard assembly will automatically be used by NuGet. Targeting individual .NET implementations increases the *.nupkg size for no benefit.
```

@drewnoakes, what do you think?

**Comment by virzak** [comment]
@samhocevar rather than cherry-picking stuff on the master branch, you could make commits into my forked repository. Besides, I just committed the style changes you asked for.

**Comment by virzak** [comment]
> @samhocevar thanks for clarifying. Perhaps give that feedback to @virzak who clearly went to a lot of effort to submit this, and give them the opportunity to have this merged. Your project though.

What he said. b73a63b is not a good etiquette.

File: Emoji.Wpf/Emoji.Wpf.csproj (hunk 3 above)
Comment (virzak) [reply Emoji.Wpf/Emoji.Wpf.csproj]: @samhocevar are these tests checked in? Everything builds for me if you have `Sdk="Microsoft.NET.Sdk.WindowsDesktop"`

**Comment by samhocevar** [comment]
> > @samhocevar thanks for clarifying. Perhaps give that feedback to @virzak who clearly went to a lot of effort to submit this, and give them the opportunity to have this merged. Your project though.
> 
> What he said. [b73a63b](https://github.com/samhocevar/emoji.wpf/commit/b73a63b6774f7a2ff0f5231195d021991d043027) is not a good etiquette.

Sorry about that, I thought that merging your branch and removing/amending the changes that I didn’t want was the best way to both preserve your authorship, keep your branch as a parent of the commit, and have a clean diff between eec295eb1 and b73a63b67. I’m not familiar with the etiquette, so please tell me if there is a better way.

Anyhow, your work is much appreciated, my apologies if it appeared otherwise.

**Comment by virzak** [comment]
@samhocevar totally understand. 

A better way would be for either you or I to edit my fork until you are satisfied and then merge or squash merge it.

You can delete your latest commit, so there are no conflicts and commit any changes into my fork. To do that you would have to `git reset --hard  eec295e` and `git push -f`

Since we were making changes in parallel they diverged quite a bit, so working on the same fork prevents that as well.

**Comment by samhocevar** [comment]
I force-pushed a clean merge commit and I think all is fine now! Thanks again.

**Comment by virzak** [comment]
@samhocevar Thank you for the library as well!

**Pull request merged**